

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>geneviz.tracks.gene &mdash; Geneviz 0.1.0.dev0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Geneviz 0.1.0.dev0 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> Geneviz
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">Geneviz</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributing</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">Geneviz</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      
    <li>geneviz.tracks.gene</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for geneviz.tracks.gene</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;This module provides basic tracks and supporting actor classes</span>
<span class="sd">for plotting gene/transcript features. The module provides the</span>
<span class="sd">following tracks:</span>

<span class="sd">    - GeneTrack: Basic track for plotting genes/transcripts from</span>
<span class="sd">        a GtfFile or GtfFrame instance on a genomic axis.</span>
<span class="sd">    - BiomartGeneTrack: Gene track that downloads its annotation</span>
<span class="sd">        directly from Biomart, thus avoiding having to provide</span>
<span class="sd">        an annotation file beforehand.</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pybiomart</span>
<span class="kn">import</span> <span class="nn">toolz</span>

<span class="kn">from</span> <span class="nn">geneviz.tracks.base</span> <span class="kn">import</span> <span class="n">Track</span>
<span class="kn">from</span> <span class="nn">geneviz.util.genomic</span> <span class="kn">import</span> <span class="n">merge_intervals</span>
<span class="kn">from</span> <span class="nn">geneviz.util.tabix</span> <span class="kn">import</span> <span class="n">GtfIterator</span>

<span class="kn">from</span> <span class="nn">.feature</span> <span class="kn">import</span> <span class="n">FeatureTrack</span>


<span class="k">class</span> <span class="nc">_BaseGeneTrack</span><span class="p">(</span><span class="n">Track</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">gene_id</span><span class="o">=</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span>
                 <span class="n">transcript_id</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">,</span>
                 <span class="nb">filter</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">collapse</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">hue</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">hue_order</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">palette</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">height</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
                 <span class="n">spacing</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
                 <span class="n">label_kws</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">__init__</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_gene_id</span> <span class="o">=</span> <span class="n">gene_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_transcript_id</span> <span class="o">=</span> <span class="n">transcript_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_collapse</span> <span class="o">=</span> <span class="n">collapse</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_filter</span> <span class="o">=</span> <span class="nb">filter</span>

        <span class="c1"># Determine which grouping feature to use.</span>
        <span class="n">collapse_group_map</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;transcript&#39;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">,</span>
            <span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">,</span>
            <span class="bp">None</span><span class="p">:</span> <span class="n">transcript_id</span>
        <span class="p">}</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">collapse_group_map</span><span class="p">[</span><span class="n">collapse</span><span class="p">]</span>

        <span class="c1"># Setup kws for feature track.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_track_kws</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;hue&#39;</span><span class="p">:</span> <span class="n">hue</span><span class="p">,</span>
            <span class="s1">&#39;hue_order&#39;</span><span class="p">:</span> <span class="n">hue_order</span><span class="p">,</span>
            <span class="s1">&#39;palette&#39;</span><span class="p">:</span> <span class="n">palette</span><span class="p">,</span>
            <span class="s1">&#39;height&#39;</span><span class="p">:</span> <span class="n">height</span><span class="p">,</span>
            <span class="s1">&#39;spacing&#39;</span><span class="p">:</span> <span class="n">spacing</span><span class="p">,</span>
            <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="n">group</span><span class="p">,</span>
            <span class="s1">&#39;group&#39;</span><span class="p">:</span> <span class="n">group</span><span class="p">,</span>
            <span class="s1">&#39;strand_junctions&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">,</span>
            <span class="s1">&#39;label_kws&#39;</span><span class="p">:</span> <span class="n">label_kws</span>
        <span class="p">}</span>

    <span class="k">def</span> <span class="nf">get_height</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">region</span><span class="p">,</span> <span class="n">ax</span><span class="p">):</span>
        <span class="n">track</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_build_track</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">track</span><span class="o">.</span><span class="n">get_height</span><span class="p">(</span><span class="n">region</span><span class="p">,</span> <span class="n">ax</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_build_track</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">region</span><span class="p">):</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fetch_exons</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">FeatureTrack</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">_track_kws</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_fetch_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">region</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_fetch_exons</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">region</span><span class="p">):</span>
        <span class="n">exons</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fetch_data</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_filter</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">exons</span> <span class="o">=</span> <span class="n">exons</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_filter</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collapse</span> <span class="o">==</span> <span class="s1">&#39;gene&#39;</span><span class="p">:</span>
            <span class="n">exons</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collapse_gene</span><span class="p">(</span><span class="n">exons</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collapse</span> <span class="o">==</span> <span class="s1">&#39;transcript&#39;</span><span class="p">:</span>
            <span class="n">exons</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collapse_transcripts</span><span class="p">(</span><span class="n">exons</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collapse</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Unexpected value for collapse&#39;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">exons</span>

    <span class="k">def</span> <span class="nf">_collapse_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exons</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Collapses exons into a single gene body.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">exons</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_gene_id</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span>
            <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="s1">&#39;first&#39;</span><span class="p">,</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span>
            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="s1">&#39;first&#39;</span>
        <span class="p">})</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_collapse_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exons</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Collapses multiple transcripts into a single transcript.&quot;&quot;&quot;</span>

        <span class="k">def</span> <span class="nf">_collapse_grp</span><span class="p">(</span><span class="n">grp</span><span class="p">):</span>
            <span class="n">first</span> <span class="o">=</span> <span class="n">grp</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">starts</span><span class="p">,</span> <span class="n">ends</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">merge_intervals</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">grp</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">grp</span><span class="o">.</span><span class="n">end</span><span class="p">)))</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
                <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span>
                <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">starts</span><span class="p">,</span>
                <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="n">ends</span><span class="p">,</span>
                <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_gene_id</span><span class="p">:</span> <span class="n">first</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_gene_id</span><span class="p">]</span>
            <span class="p">})</span>

        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">(</span><span class="n">_collapse_grp</span><span class="p">(</span><span class="n">grp</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">grp</span> <span class="ow">in</span> <span class="n">exons</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_gene_id</span><span class="p">)),</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">ignore_index</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">draw</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">region</span><span class="p">,</span> <span class="n">ax</span><span class="p">):</span>
        <span class="n">track</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_build_track</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">track</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">region</span><span class="p">,</span> <span class="n">ax</span><span class="p">)</span>


<div class="viewcode-block" id="GtfTrack"><a class="viewcode-back" href="../../../api.html#geneviz.tracks.GtfTrack">[docs]</a><span class="k">class</span> <span class="nc">GtfTrack</span><span class="p">(</span><span class="n">_BaseGeneTrack</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">gtf_path</span><span class="p">,</span>
                 <span class="n">gene_id</span><span class="o">=</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span>
                 <span class="n">transcript_id</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">,</span>
                 <span class="n">collapse</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="nb">filter</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">hue</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">hue_order</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">palette</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">height</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
                 <span class="n">spacing</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
                 <span class="n">label_kws</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span>
            <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
            <span class="n">transcript_id</span><span class="o">=</span><span class="n">transcript_id</span><span class="p">,</span>
            <span class="n">collapse</span><span class="o">=</span><span class="n">collapse</span><span class="p">,</span>
            <span class="nb">filter</span><span class="o">=</span><span class="nb">filter</span><span class="p">,</span>
            <span class="n">hue</span><span class="o">=</span><span class="n">hue</span><span class="p">,</span>
            <span class="n">hue_order</span><span class="o">=</span><span class="n">hue_order</span><span class="p">,</span>
            <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>
            <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span>
            <span class="n">spacing</span><span class="o">=</span><span class="n">spacing</span><span class="p">,</span>
            <span class="n">label_kws</span><span class="o">=</span><span class="n">label_kws</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_gtf_path</span> <span class="o">=</span> <span class="n">gtf_path</span>

    <span class="k">def</span> <span class="nf">_fetch_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">region</span><span class="p">):</span>
        <span class="c1"># TODO: Ensure we fetch the full transcript.</span>

        <span class="c1"># Fetch exon records from gtf.</span>
        <span class="n">gtf_iter</span> <span class="o">=</span> <span class="n">GtfIterator</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_gtf_path</span><span class="p">)</span>
        <span class="n">records</span> <span class="o">=</span> <span class="n">gtf_iter</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="o">*</span><span class="n">region</span><span class="p">)</span>

        <span class="c1"># Convert to DataFrame.</span>
        <span class="n">exons</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_records_to_exons</span><span class="p">(</span><span class="n">records</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">exons</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">_records_to_exons</span><span class="p">(</span><span class="n">cls</span><span class="p">,</span> <span class="n">records</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">feature</span> <span class="o">==</span> <span class="s1">&#39;exon&#39;</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">cls</span><span class="o">.</span><span class="n">_parse_exon_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_parse_exon_record</span><span class="p">(</span><span class="n">record</span><span class="p">):</span>
        <span class="n">basic_info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="n">record</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">record</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="n">record</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="n">numeric_strand</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="k">return</span> <span class="n">toolz</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">basic_info</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">record</span><span class="p">))</span></div>


<span class="k">def</span> <span class="nf">numeric_strand</span><span class="p">(</span><span class="n">strand</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Converts the +/- strand notation to a numeric 1/-1 format.&quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">strand</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">}</span>
    <span class="k">return</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="o">-</span><span class="mi">1</span>


<div class="viewcode-block" id="BiomartTrack"><a class="viewcode-back" href="../../../api.html#geneviz.tracks.BiomartTrack">[docs]</a><span class="k">class</span> <span class="nc">BiomartTrack</span><span class="p">(</span><span class="n">_BaseGeneTrack</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Track for plotting gene/transcript annotations from Biomart.</span>

<span class="sd">    Gene draws genes/transcripts on a genomic axis, in a stacked</span>
<span class="sd">    fashion if transcripts/genes are overlapping. Gene annotations</span>
<span class="sd">    are queried directly from Biomart.</span>

<span class="sd">    Args:</span>
<span class="sd">        host (str): Biomart host address.</span>
<span class="sd">        mart (str): Biomart mart name.</span>
<span class="sd">        dataset (str): Biomart dataset name.</span>
<span class="sd">        **kwargs: Keywords are passed to the GeneTrack constructor,</span>
<span class="sd">            which is used internally to draw transcripts/genes.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">host</span><span class="o">=</span><span class="s1">&#39;http://www.ensembl.org&#39;</span><span class="p">,</span>
                 <span class="n">mart</span><span class="o">=</span><span class="s1">&#39;ENSEMBL_MART_ENSEMBL&#39;</span><span class="p">,</span>
                 <span class="n">dataset</span><span class="o">=</span><span class="s1">&#39;hsapiens_gene_ensembl&#39;</span><span class="p">,</span>
                 <span class="n">gene_id</span><span class="o">=</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span>
                 <span class="n">transcript_id</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">,</span>
                 <span class="n">collapse</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="nb">filter</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">hue</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">hue_order</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">palette</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">height</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
                 <span class="n">spacing</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
                 <span class="n">label_kws</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span>
            <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
            <span class="n">transcript_id</span><span class="o">=</span><span class="n">transcript_id</span><span class="p">,</span>
            <span class="n">collapse</span><span class="o">=</span><span class="n">collapse</span><span class="p">,</span>
            <span class="nb">filter</span><span class="o">=</span><span class="nb">filter</span><span class="p">,</span>
            <span class="n">hue</span><span class="o">=</span><span class="n">hue</span><span class="p">,</span>
            <span class="n">hue_order</span><span class="o">=</span><span class="n">hue_order</span><span class="p">,</span>
            <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>
            <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span>
            <span class="n">spacing</span><span class="o">=</span><span class="n">spacing</span><span class="p">,</span>
            <span class="n">label_kws</span><span class="o">=</span><span class="n">label_kws</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">pybiomart</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Pybiomart must be installed to use &#39;</span>
                             <span class="s1">&#39;the BiomartGeneTrack &#39;</span><span class="p">)</span>

        <span class="c1"># Fetch dataset from server.</span>
        <span class="n">server</span> <span class="o">=</span> <span class="n">pybiomart</span><span class="o">.</span><span class="n">Server</span><span class="p">(</span><span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_dataset</span> <span class="o">=</span> <span class="p">(</span><span class="n">server</span><span class="o">.</span><span class="n">marts</span><span class="p">[</span><span class="n">mart</span><span class="p">]</span><span class="o">.</span><span class="n">datasets</span><span class="p">[</span><span class="n">dataset</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">_fetch_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">region</span><span class="p">):</span>
        <span class="c1"># TODO: Fetch transcript name instead of id (needs extra query).</span>

        <span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">region</span>

        <span class="c1"># Determine which transcripts are within region.</span>
        <span class="n">transcripts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_dataset</span><span class="o">.</span><span class="n">query</span><span class="p">(</span>
            <span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ensembl_transcript_id&#39;</span><span class="p">],</span>
            <span class="n">filters</span><span class="o">=</span><span class="p">{</span>
                <span class="s1">&#39;chromosome_name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">chromosome</span><span class="p">],</span>
                <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">start</span><span class="p">],</span>
                <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">end</span><span class="p">]</span>
            <span class="p">})</span>
        <span class="n">transcript_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">transcripts</span><span class="p">[</span><span class="n">transcripts</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>

        <span class="c1"># Retrieve exons for these transcripts.</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_dataset</span><span class="o">.</span><span class="n">query</span><span class="p">(</span>
            <span class="n">attributes</span><span class="o">=</span><span class="p">[</span>
                <span class="s1">&#39;external_gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;ensembl_gene_id&#39;</span><span class="p">,</span>
                <span class="s1">&#39;ensembl_transcript_id&#39;</span><span class="p">,</span> <span class="s1">&#39;exon_chrom_start&#39;</span><span class="p">,</span> <span class="s1">&#39;exon_chrom_end&#39;</span><span class="p">,</span>
                <span class="s1">&#39;strand&#39;</span>
            <span class="p">],</span>
            <span class="n">filters</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;link_ensembl_transcript_stable_id&#39;</span><span class="p">:</span> <span class="n">transcript_ids</span><span class="p">},</span>
            <span class="n">use_attr_names</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

        <span class="c1"># Reshape data to conform to expected format.</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="n">chromosome</span><span class="p">,</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;exon_chrom_start&#39;</span><span class="p">],</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;exon_chrom_end&#39;</span><span class="p">],</span>
            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">],</span>
            <span class="s1">&#39;gene_name&#39;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;external_gene_name&#39;</span><span class="p">],</span>
            <span class="s1">&#39;gene_id&#39;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;ensembl_gene_id&#39;</span><span class="p">],</span>
            <span class="s1">&#39;transcript_id&#39;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;ensembl_transcript_id&#39;</span><span class="p">]</span>
        <span class="p">})</span>

        <span class="k">return</span> <span class="n">data</span></div>
</pre></div>

           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.1.0.dev0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>